lib/java/com/google/android/material/internal/CollapsingTextHelper.java

```diff
@@ -691,13 +783,56 @@ private void calculateUsingTextSize(final float textSize) {
       // Use linear text scaling if we're scaling the canvas
       textPaint.setLinearText(scale != 1f);
 
-      // If we don't currently have text to draw, or the text size has changed, ellipsize...
-      final CharSequence title =
-          TextUtils.ellipsize(text, textPaint, availableWidth, TextUtils.TruncateAt.END);
-      if (!TextUtils.equals(title, textToDraw)) {
-        textToDraw = title;
+      StaticLayout layout = new StaticLayout(text, textPaint, (int) availableWidth,
+          Layout.Alignment.ALIGN_NORMAL, lineSpacingMultiplier, lineSpacingExtra, false);
+      CharSequence truncatedText;
+      if (layout.getLineCount() > maxLines) {
+        int lastLine = maxLines - 1;
+        CharSequence textBefore =
+            lastLine > 0 ? text.subSequence(0, layout.getLineEnd(lastLine - 1)) : "";
+        CharSequence lineText = text.subSequence(layout.getLineStart(lastLine),
+            layout.getLineEnd(lastLine));
+        // if last char in line is space, move it behind the ellipsis
+        CharSequence lineEnd = "";
+        if (lineText.charAt(lineText.length() - 1) == ' ') {
+          lineEnd = lineText.subSequence(lineText.length() - 1, lineText.length());
+          lineText = lineText.subSequence(0, lineText.length() - 1);
+        }
+        // insert ellipsis character
+        lineText = TextUtils.concat(lineText, "\u2026", lineEnd);
+        // if the text is too long, truncate it
+        CharSequence truncatedLineText = TextUtils.ellipsize(lineText, textPaint,
+            availableWidth, TextUtils.TruncateAt.END);
+        truncatedText = TextUtils.concat(textBefore, truncatedLineText);
+
+      } else {
+        truncatedText = text;
```

Comment:
In this case should it potentially be ellipsized?

Reply:
Umm, the `else` case is when the number of lines required to show the whole title (`layout.getLineCount()`) is `<= maxLines` - so the whole text can be shown. Why should it be ellipsized then?

Reply:
yeah think I turned that logic around in my head! Looks right.